tests/integration/test_core.py

```diff
@@ -540,3 +545,111 @@ async def test_openid_plugin(
+):
+    """
+    arrange: after WordPress charm has been deployed and relations established.
+    act: loki charm joins relation
+    assert: loki joins relation successfully, logs are being output to container and to files for
+        loki to scrape.
+    """
+    assert ops_test.model
+    loki: Application = await ops_test.model.deploy("loki-k8s", channel="stable", trust=True)
+
+    await ops_test.model.relate(application_name, loki.name)
+    await ops_test.model.wait_for_idle(apps=[application_name, loki.name], status="active")
+
+    status: FullStatus = await ops_test.model.get_status(filters=[loki.name])
+    for unit in status.applications[loki.name].units.values():
+        series = requests.get(f"http://{unit.address}:3100/loki/api/v1/series", timeout=10).json()
+        log_files = tuple(series_data["filename"] for series_data in series["data"])
+        assert "/var/log/apache2/error.log" in log_files
+        assert "/var/log/apache2/access.log" in log_files
+        log_query = requests.get(
+            f"http://{unit.address}:3100/loki/api/v1/query",
+            timeout=10,
+            params={"query": f'{{juju_application="{application_name}"}}'},
+        ).json()
+        assert len(log_query["data"]["result"])
+    kube_log = kube_core_client.read_namespaced_pod_log(
+        name=f"{application_name}-0", namespace=ops_test.model_name, container="wordpress"
+    )
+    assert kube_log
+
+
+async def test_grafana_integration(
+    ops_test: pytest_operator.plugin.OpsTest,
+    application_name: str,
+):
+    """
+    arrange: after WordPress charm has been deployed and relations established among cos.
+    act: grafana charm joins relation
+    assert: grafana wordpress dashboard can be found
+    """
+    assert ops_test.model
+    grafana: Application = await ops_test.model.deploy("grafana-k8s", trust=True)
+
+    await ops_test.model.relate(application_name, grafana.name)
+    await ops_test.model.relate("prometheus-k8s:grafana-source", f"{grafana.name}:grafana-source")
+    await ops_test.model.relate("loki-k8s:grafana-source", f"{grafana.name}:grafana-source")
+    await ops_test.model.wait_for_idle(apps=[application_name, grafana.name], status="active")
+
+    action: Action = await grafana.units[0].run_action("get-admin-password")
+    await action.wait()
+    password = action.results["admin-password"]
+    status: FullStatus = await ops_test.model.get_status(filters=[grafana.name])
+    for unit in status.applications[grafana.name].units.values():
+        sess = requests.session()
+        sess.post(
+            f"http://{unit.address}:3000/login",
+            json={
+                "user": "admin",
+                "password": password,
+            },
+        ).raise_for_status()
+        datasources = sess.get(f"http://{unit.address}:3000/api/datasources", timeout=10).json()
+        datasource_types = tuple(datasource["type"] for datasource in datasources)
```

Comment:
Similar to above, should this be a set given the next lines check for elements within the collection?